Fixes thermomachine power bug

AlinaStarkova opened the pull request:
if it has no power, its hatch is open, or is broken, it will turn off and set power usage to 0

<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
fixes #56257 .... Gilkher when he made thermomachines he didnt add in not a single check to make sure the machine actually *had* power to use, so was able to be used in space, with no power, so long as the machine was on. I simply added the proc that processed its gas contents behind a simple check to ensure it has power, isnt broken, and the maint cover is closed. if one of those is true, it will turn the machine back off and set power usage to 0

Before the fix:
https://user-images.githubusercontent.com/40489693/104986520-28165d00-59e1-11eb-9402-a883baa61dea.mp4


After the fix:
https://i.gyazo.com/3e3392095797a8ffc78178504117be8e.mp4


<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

fix bug, unga
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: Nari Harimoto
fix: fixed thermomachines still working even without power, apc off, in space etc...
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

AlinaStarkova committed:
checks if the machine actually has power

if it has no power, its hatch is open, or is broken, it will turn off and set power usage to 0

Jared-Fogle commented on code/modules/atmospherics/machinery/components/unary_devices/thermomachine.dm:
```diff
@@ -109,25 +109,32 @@
-	else
-		active_power_usage = idle_power_usage
-	return 1
+	return
```
Remove the trailing return--it is implicit.

Jared-Fogle commented on code/modules/atmospherics/machinery/components/unary_devices/thermomachine.dm:
```diff
@@ -109,25 +109,32 @@
+			var/combined_energy = heat_capacity * target_temperature + air_heat_capacity * air_contents.temperature
+			air_contents.temperature = combined_energy/combined_heat_capacity
+
+		var/temperature_delta= abs(old_temperature - air_contents.temperature)
```
```suggestion
		var/temperature_delta = abs(old_temperature - air_contents.temperature)
```

Jared-Fogle commented on code/modules/atmospherics/machinery/components/unary_devices/thermomachine.dm:
```diff
@@ -109,25 +109,32 @@
+			update_parents()
+		else
+			active_power_usage = idle_power_usage
+		return 1
```
What's the 1 for? Is it in place of `TRUE`, or is this meant to be a number?

AlinaStarkova replied:
it was there before hand, the whole procs parent defaults to FALSE, setting to true actually shuts it off from what i can tell. in any case, i refuse to touch the code for atmos, its outside the scope of this PR, im only here to fix the power issue

Jared-Fogle replied:
I'm just asking you to change it to `return TRUE` while you're here 😛

Jared-Fogle commented on code/modules/atmospherics/machinery/components/unary_devices/thermomachine.dm:
```diff
@@ -109,25 +109,32 @@
+		else
+			active_power_usage = idle_power_usage
+		return 1
+	else if(!is_operational)
```
Should this not just be `else`?

LemonInTheDark replied:
Should really be doing this in an override of `set_is_operational`, check is_operational for the on/off toggles to be clean about things.

LemonInTheDark replied:
As another option, add a check for is_operational in the same places we check for on, so the machine will come back to life when the power comes back

LemonInTheDark commented on code/modules/atmospherics/machinery/components/unary_devices/thermomachine.dm:
```diff
@@ -109,25 +109,32 @@
 /obj/machinery/atmospherics/components/unary/thermomachine/process_atmos()
 	..()
-	if(!on || !nodes[1])
+	if(is_operational)
```
The var itself should be added to the !on || !nodes[1] bit

AlinaStarkova replied:
i dont know, i didnt want to mess with the code itself because i know literally nothing about atmos, i was just trying to fix the bug without trying to mess with any code related to atmos. i can try it tomorrow to see if that works

LemonInTheDark replied:
atmos isn't all that spooky, if you need help feel free to ask tho, I get why it might be worrying

AlinaStarkova committed:
Update thermomachine.dm

checks if it was on before it lost power, if yes, it turns back on when power turns on

simpler method of checking if it lost power

AlinaStarkova committed:
added a space, doesnt affect code

AlinaStarkova committed:
puts the last return where it should be

Jared-Fogle commented on code/modules/atmospherics/machinery/components/unary_devices/thermomachine.dm:
```diff
@@ -196,6 +199,10 @@
 			use_power = on ? ACTIVE_POWER_USE : IDLE_POWER_USE
 			investigate_log("was turned [on ? "on" : "off"] by [key_name(usr)]", INVESTIGATE_ATMOS)
 			. = TRUE
+			if(on == TRUE)
+				was_on = TRUE   //if the machine was manually turned on, ensure it remembers it
+			else
+				was_on = FALSE  //else dont remember it was on, because it wasnt.
```
```suggestion
			was_on = on
```

AlinaStarkova replied:
that wont work though, because then hitting the button to turn it off will also set it to was_on, when it actually needs to set it back to false when turned off

Jared-Fogle replied:
This is exactly the same as your code, unless on is for whatever reason not always TRUE/FALSE.

Please recheck your code.

AlinaStarkova replied:
changed to `was_on = !was_on`

AlinaStarkova committed:
changed how it checks if it was on

AlinaStarkova committed:
Merge branch 'master' into fix-thermomachine
